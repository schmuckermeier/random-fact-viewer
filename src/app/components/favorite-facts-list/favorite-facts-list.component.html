<table [dataSource]="favoriteFacts$" class="mat-elevation-z8" mat-table>

  <tr *matHeaderRowDef="displayedColumns" mat-header-row></tr>
  <tr *matRowDef="let row; columns: displayedColumns;" mat-row></tr>

  <ng-container matColumnDef="name">
    <th *matHeaderCellDef mat-header-cell>Random Fact</th>
    <td *matCellDef="let randomFact" mat-cell> {{ randomFact.text }}</td>
  </ng-container>

  <ng-container matColumnDef="source">
    <th *matHeaderCellDef mat-header-cell>Fact Source</th>
    <td *matCellDef="let randomFact" mat-cell> {{ randomFact.source }}</td>
  </ng-container>

  <ng-container matColumnDef="delete">
    <th *matHeaderCellDef mat-header-cell></th>
    <td *matCellDef="let randomFact" mat-cell>
      <button (click)="removeFromFavorites(randomFact)" mat-mini-fab>
        <mat-icon>delete</mat-icon>
      </button>
    </td>
  </ng-container>
</table>
