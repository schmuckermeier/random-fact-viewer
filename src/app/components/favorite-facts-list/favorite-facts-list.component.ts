import {Component, inject} from '@angular/core';
import {MatIconModule} from '@angular/material/icon';
import {RandomFact} from '../../services/random-fact';
import {FavoriteFactsService} from '../../services/favorite-facts.service';
import {MatButtonModule} from '@angular/material/button';
import {MatTableModule} from '@angular/material/table';

@Component({
  selector: 'app-favorite-facts-list',
  imports: [MatIconModule, MatButtonModule, MatTableModule],
  templateUrl: './favorite-facts-list.component.html',
  styleUrl: './favorite-facts-list.component.css'
})
export class FavoriteFactsListComponent {
  displayedColumns: string[] = ['name', 'source', 'delete'];

  private favoriteFactsService = inject(FavoriteFactsService);

  favoriteFacts$ = this.favoriteFactsService.getFacts();

  removeFromFavorites(fact: RandomFact) {
    this.favoriteFactsService.removeFact(fact);
  }
}
