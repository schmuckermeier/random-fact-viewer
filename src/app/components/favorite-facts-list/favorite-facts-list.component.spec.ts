import { ComponentFixture, TestBed } from '@angular/core/testing';
import { FavoriteFactsListComponent } from './favorite-facts-list.component';
import {FavoriteFactsService} from '../../services/favorite-facts.service';
import {of} from 'rxjs';
import {mockRandomFacts} from '../../testing/mockRandomFacts';
import {queryById} from '../../testing/testing-utils';

describe('FavoriteFactsListComponent', () => {
  let fixture: ComponentFixture<FavoriteFactsListComponent>;

  let mockFavoriteFactsService = {
    getFacts: jest.fn(() => of(mockRandomFacts)),
    removeFact: jest.fn()
  };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [FavoriteFactsListComponent],
      providers: [{provide: FavoriteFactsService, useValue: mockFavoriteFactsService}]
    })
    .compileComponents();

    fixture = TestBed.createComponent(FavoriteFactsListComponent);
    fixture.detectChanges();
  });

  it('should render with initial data in table', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should remove favorite fact on click', () => {
    const firstRandomFact = mockRandomFacts[0];
    const deleteButton = queryById(fixture, "delete-" + firstRandomFact.id)
    deleteButton.nativeElement.click();

    expect(mockFavoriteFactsService.removeFact).toHaveBeenCalledWith(firstRandomFact);
  })
});
