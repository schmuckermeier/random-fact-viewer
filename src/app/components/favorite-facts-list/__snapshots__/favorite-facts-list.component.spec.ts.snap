// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteFactsListComponent should render with initial data in table 1`] = `
<app-favorite-facts-list
  displayedColumns={[Function Array]}
  favoriteFacts$={[Function Observable]}
  favoriteFactsService={[Function Object]}
>
  <table
    class="mat-mdc-table mdc-data-table__table cdk-table mat-elevation-z8"
    mat-table=""
    role="table"
  >
    <thead
      role="rowgroup"
    >
      <tr
        class="mat-mdc-header-row mdc-data-table__header-row cdk-header-row"
        mat-header-row=""
        role="row"
      >
        <th
          class="mat-mdc-header-cell mdc-data-table__header-cell cdk-header-cell cdk-column-name mat-column-name"
          mat-header-cell=""
          role="columnheader"
        >
          Random Fact
        </th>
        <th
          class="mat-mdc-header-cell mdc-data-table__header-cell cdk-header-cell cdk-column-source mat-column-source"
          mat-header-cell=""
          role="columnheader"
        >
          Fact Source
        </th>
        <th
          class="mat-mdc-header-cell mdc-data-table__header-cell cdk-header-cell cdk-column-delete mat-column-delete"
          mat-header-cell=""
          role="columnheader"
        />
      </tr>
    </thead>
    <tbody
      class="mdc-data-table__content"
      role="rowgroup"
    >
      <tr
        class="mat-mdc-row mdc-data-table__row cdk-row"
        mat-row=""
        role="row"
      >
        <td
          class="mat-mdc-cell mdc-data-table__cell cdk-cell cdk-column-name mat-column-name"
          mat-cell=""
          role="cell"
        >
           Golf courses cover 4% of North America.
        </td>
        <td
          class="mat-mdc-cell mdc-data-table__cell cdk-cell cdk-column-source mat-column-source"
          mat-cell=""
          role="cell"
        >
           djtech.net
        </td>
        <td
          class="mat-mdc-cell mdc-data-table__cell cdk-cell cdk-column-delete mat-column-delete"
          mat-cell=""
          role="cell"
        >
          <button
            class="mdc-fab mat-mdc-fab-base mdc-fab--mini mat-mdc-mini-fab mat-accent mat-mdc-button-base"
            id="delete-78e7f1adff152ae7ad66821de0ca076b"
            mat-mini-fab=""
            mat-ripple-loader-class-name="mat-mdc-button-ripple"
            mat-ripple-loader-uninitialized=""
          >
            <span
              class="mat-mdc-button-persistent-ripple mdc-fab__ripple"
            />
            <mat-icon
              aria-hidden="true"
              class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
              data-mat-icon-type="font"
              role="img"
            >
              delete
            </mat-icon>
            <span
              class="mdc-button__label"
            />
            <span
              class="mat-focus-indicator"
            />
            <span
              class="mat-mdc-button-touch-target"
            />
          </button>
        </td>
      </tr>
      <tr
        class="mat-mdc-row mdc-data-table__row cdk-row"
        mat-row=""
        role="row"
      >
        <td
          class="mat-mdc-cell mdc-data-table__cell cdk-cell cdk-column-name mat-column-name"
          mat-cell=""
          role="cell"
        >
           Because of the rotation of the earth, an object can be thrown farther if it is thrown west.
        </td>
        <td
          class="mat-mdc-cell mdc-data-table__cell cdk-cell cdk-column-source mat-column-source"
          mat-cell=""
          role="cell"
        >
           djtech.net
        </td>
        <td
          class="mat-mdc-cell mdc-data-table__cell cdk-cell cdk-column-delete mat-column-delete"
          mat-cell=""
          role="cell"
        >
          <button
            class="mdc-fab mat-mdc-fab-base mdc-fab--mini mat-mdc-mini-fab mat-accent mat-mdc-button-base"
            id="delete-f21f3b1d3923002b864c516762575d6b"
            mat-mini-fab=""
            mat-ripple-loader-class-name="mat-mdc-button-ripple"
            mat-ripple-loader-uninitialized=""
          >
            <span
              class="mat-mdc-button-persistent-ripple mdc-fab__ripple"
            />
            <mat-icon
              aria-hidden="true"
              class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
              data-mat-icon-type="font"
              role="img"
            >
              delete
            </mat-icon>
            <span
              class="mdc-button__label"
            />
            <span
              class="mat-focus-indicator"
            />
            <span
              class="mat-mdc-button-touch-target"
            />
          </button>
        </td>
      </tr>
      <tr
        class="mat-mdc-row mdc-data-table__row cdk-row"
        mat-row=""
        role="row"
      >
        <td
          class="mat-mdc-cell mdc-data-table__cell cdk-cell cdk-column-name mat-column-name"
          mat-cell=""
          role="cell"
        >
           In medieval France, unfaithful wives were made to chase a chicken through town naked.
        </td>
        <td
          class="mat-mdc-cell mdc-data-table__cell cdk-cell cdk-column-source mat-column-source"
          mat-cell=""
          role="cell"
        >
           djtech.net
        </td>
        <td
          class="mat-mdc-cell mdc-data-table__cell cdk-cell cdk-column-delete mat-column-delete"
          mat-cell=""
          role="cell"
        >
          <button
            class="mdc-fab mat-mdc-fab-base mdc-fab--mini mat-mdc-mini-fab mat-accent mat-mdc-button-base"
            id="delete-6064df56adfae9cdca9cc428312343be"
            mat-mini-fab=""
            mat-ripple-loader-class-name="mat-mdc-button-ripple"
            mat-ripple-loader-uninitialized=""
          >
            <span
              class="mat-mdc-button-persistent-ripple mdc-fab__ripple"
            />
            <mat-icon
              aria-hidden="true"
              class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
              data-mat-icon-type="font"
              role="img"
            >
              delete
            </mat-icon>
            <span
              class="mdc-button__label"
            />
            <span
              class="mat-focus-indicator"
            />
            <span
              class="mat-mdc-button-touch-target"
            />
          </button>
        </td>
      </tr>
    </tbody>
    <tfoot
      class="mat-mdc-table-sticky"
      role="rowgroup"
      style="display: none; bottom: 0px; z-index: 10;"
    >
    </tfoot>
  </table>
</app-favorite-facts-list>
`;
